// @dydxprotocol/starkex-eth/build/src/modules/CollateralToken.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollateralToken = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const Constants_1 = require("../lib/Constants");
const ContractCallHelpers_1 = require("../lib/ContractCallHelpers");
class CollateralToken {
    constructor(contracts) {
        this.contracts = contracts;
        this.token = contracts.collateralToken;
    }
    getAddress() {
        return this.contracts.collateralToken.options.address;
    }
    getAssetId() {
        return Constants_1.COLLATERAL_ASSET_ID[this.contracts.networkId];
    }
    async getAllowance({ ownerAddress, spenderAddress, }, options) {
        const allowance = await this.contracts.call(this.token.methods.allowance(ownerAddress, spenderAddress), options);
        return (0, ContractCallHelpers_1.uint256ToHumanCollateralTokenAmount)(allowance);
    }
    async getBalance({ ownerAddress, }, options) {
        const balance = await this.contracts.call(this.token.methods.balanceOf(ownerAddress), options);
        return (0, ContractCallHelpers_1.uint256ToHumanCollateralTokenAmount)(balance);
    }
    async getTotalSupply(options) {
        const supply = await this.contracts.call(this.token.methods.totalSupply(), options);
        return (0, ContractCallHelpers_1.uint256ToHumanCollateralTokenAmount)(supply);
    }
    async getName(options) {
        return this.contracts.call(this.token.methods.name(), options);
    }
    async getSymbol(options) {
        return this.contracts.call(this.token.methods.symbol(), options);
    }
    async getDecimals(options) {
        const decimals = await this.contracts.call(this.token.methods.decimals(), options);
        return new bignumber_js_1.default(decimals).toNumber();
    }
    async getExchangeAllowance({ ownerAddress, }, options) {
        return this.getAllowance({
            ownerAddress,
            spenderAddress: this.contracts.starkwarePerpetual.options.address,
        }, options);
    }
    async setAllowance({ spenderAddress, humanAmount, }, options) {
        return this.contracts.send(this.token, this.token.methods.approve(spenderAddress, (0, ContractCallHelpers_1.humanCollateralAmountToUint256)(humanAmount)), options);
    }
    async setExchangeAllowance({ humanAmount, }, options) {
        return this.setAllowance({
            spenderAddress: this.contracts.starkwarePerpetual.options.address,
            humanAmount,
        }, options);
    }
    async setMaximumAllowance({ spenderAddress, }, options) {
        return this.contracts.send(this.token, this.token.methods.approve(spenderAddress, Constants_1.INTEGERS.ONES_255), options);
    }
    async setMaximumExchangeAllowance(options) {
        return this.setMaximumAllowance({
            spenderAddress: this.contracts.starkwarePerpetual.options.address,
        }, options);
    }
    async unsetExchangeAllowance(options) {
        return this.setAllowance({
            spenderAddress: this.contracts.starkwarePerpetual.options.address,
            humanAmount: '0',
        }, options);
    }
    async transfer({ toAddress, humanAmount, }, options) {
        return this.contracts.send(this.token, this.token.methods.transfer(toAddress, (0, ContractCallHelpers_1.humanCollateralAmountToUint256)(humanAmount)), options);
    }
    async transferFrom({ fromAddress, toAddress, humanAmount, }, options) {
        return this.contracts.send(this.token, this.token.methods.transferFrom(fromAddress, toAddress, (0, ContractCallHelpers_1.humanCollateralAmountToUint256)(humanAmount)), options);
    }
    async isBlacklisted({ address, }, options) {
        const result = await this.contracts.call(this.token.methods.isBlacklisted(address), options);
        return result;
    }
}
exports.CollateralToken = CollateralToken;
