// @dydxprotocol/starkex-eth/build/src/modules/Logs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logs = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
class Logs {
    constructor(web3, contracts) {
        this.web3 = web3;
        this.contracts = contracts;
    }
    get contractsByAddress() {
        var _a, _b;
        if (!this._contractsByAddress) {
            this._contractsByAddress = {};
            for (const { contract, isTest } of this.contracts.contractsList) {
                const contractAddress = (_b = (_a = contract === null || contract === void 0 ? void 0 : contract.options) === null || _a === void 0 ? void 0 : _a.address) === null || _b === void 0 ? void 0 : _b.toLowerCase();
                if (!isTest && contractAddress) {
                    if (!this._contractsByAddress[contractAddress]) {
                        this._contractsByAddress[contractAddress] = [];
                    }
                    this._contractsByAddress[contractAddress].push(contract);
                }
            }
        }
        return this._contractsByAddress;
    }
    parseLogs(receipt) {
        if (receipt.logs) {
            const events = JSON.parse(JSON.stringify(receipt.logs));
            return events
                .map((e) => this.parseLog(e))
                .filter(notEmpty);
        }
        if (receipt.events) {
            const tempEvents = JSON.parse(JSON.stringify(receipt.events));
            const events = [];
            Object.values(tempEvents).forEach((e) => {
                if (Array.isArray(e)) {
                    e.forEach((ev) => events.push(ev));
                }
                else {
                    events.push(e);
                }
            });
            events.sort((a, b) => a.logIndex - b.logIndex);
            return events.map((e) => this.parseEvent(e)).filter(notEmpty);
        }
        throw new Error('Receipt has no logs');
    }
    parseEvent(event) {
        return this.parseLog({
            address: event.address,
            data: event.raw.data,
            topics: event.raw.topics,
            logIndex: event.logIndex,
            transactionHash: event.transactionHash,
            transactionIndex: event.transactionIndex,
            blockHash: event.blockHash,
            blockNumber: event.blockNumber,
        });
    }
    parseLog(log) {
        const logAddress = log.address.toLowerCase();
        const contracts = this.contractsByAddress[logAddress];
        if (!contracts) {
            return null;
        }
        for (let i = 0; i < contracts.length; i += 1) {
            const parsedLog = this.parseLogWithContract(contracts[i], log);
            if (parsedLog) {
                return parsedLog;
            }
        }
        return null;
    }
    parseLogWithContract(contract, log) {
        const events = contract.options.jsonInterface.filter((e) => e.type === 'event');
        const eventJson = events.find((e) => e.signature.toLowerCase() === log.topics[0].toLowerCase());
        if (!eventJson) {
            return null;
        }
        const eventArgs = this.web3.eth.abi.decodeLog(eventJson.inputs, log.data, log.topics.slice(1));
        return {
            ...log,
            name: eventJson.name,
            args: this.parseArgs(eventJson.inputs, eventArgs),
        };
    }
    parseArgs(inputs, eventArgs) {
        const parsedObject = {};
        for (const input of inputs) {
            const { name } = input;
            parsedObject[name] = this.parseValue(input, eventArgs[name]);
        }
        return parsedObject;
    }
    parseValue(input, argValue) {
        // TODO: like all other logs except LogDeposit
        if (input.type === 'uint256') {
            // returning everything as strings
            // because listener just needs to store the data
            switch (input.name) {
                case 'starkKey':
                case 'assetType':
                case 'vaultId':
                case 'nonQuantizedAmount':
                case 'quantizedAmount':
                default:
                    return argValue;
            }
        }
        if (input.type === 'address') {
            return argValue;
        }
        if (input.type === 'bool') {
            return argValue;
        }
        if (input.type.match(/^bytes[0-9]*$/)) {
            return argValue;
        }
        if (input.type.match(/^uint[0-9]*$/)) {
            return new bignumber_js_1.default(argValue);
        }
        throw new Error(`Unknown event arg type ${input.type}`);
    }
}
exports.Logs = Logs;
function notEmpty(value) {
    return value !== null && value !== undefined;
}
