// @dydxprotocol/starkex-eth/build/src/lib/Constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.USDC_ADDRESSES = exports.ADDRESSES = exports.COLLATERAL_ASSET_ID = exports.INTEGERS = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const usdc_abi_json_1 = __importDefault(require("../contracts/usdc-abi.json"));
const types_1 = require("../types");
const BytesHelper_1 = require("./BytesHelper");
exports.INTEGERS = {
    ZERO: new bignumber_js_1.default(0),
    ONE: new bignumber_js_1.default(1),
    ONES_255: new bignumber_js_1.default('115792089237316195423570985008687907853269984665640564039457584007913129639935'), // 2**256-1
};
exports.COLLATERAL_ASSET_ID = {
    [types_1.Networks.MAINNET]: (0, BytesHelper_1.getAssetId)(usdc_abi_json_1.default.networks[1].address),
    [types_1.Networks.ROPSTEN]: (0, BytesHelper_1.getAssetId)(usdc_abi_json_1.default.networks[3].address),
};
exports.ADDRESSES = {
    ZERO: '0x0000000000000000000000000000000000000000',
    DYDX_USDC_ADDRESS_ROPSTEN: '0x8707a5bf4c2842d46b31a405ba41b858c0f876c4',
    USDC_ADDRESS_MAINNET: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
};
exports.USDC_ADDRESSES = {
    [types_1.Networks.ROPSTEN]: exports.ADDRESSES.DYDX_USDC_ADDRESS_ROPSTEN,
    [types_1.Networks.MAINNET]: exports.ADDRESSES.USDC_ADDRESS_MAINNET,
};
