// @dydxprotocol/starkex-eth/build/src/lib/ContractCallHelpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uint256ToHumanTokenAmount = exports.uint256ToHumanCollateralTokenAmount = exports.uint256ToHumanEthAmount = exports.humanEthAmountToUint256 = exports.humanCollateralAmountToUint256 = exports.humanTokenAmountToUint256 = exports.bignumberableToUint256 = exports.starkKeyToUint256 = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const types_1 = require("../types");
// To-Contract Helpers
function starkKeyToUint256(starkKey) {
    return new bignumber_js_1.default(starkKey, 16).toFixed(0);
}
exports.starkKeyToUint256 = starkKeyToUint256;
function bignumberableToUint256(amount) {
    const result = new bignumber_js_1.default(amount);
    // Solidity only takes integers.
    if (!result.isInteger()) {
        throw new Error(`Amount cannot be used in contract call: ${result.toFixed()}`);
    }
    return result.toFixed(0);
}
exports.bignumberableToUint256 = bignumberableToUint256;
function humanTokenAmountToUint256(humanAmount, decimals) {
    return bignumberableToUint256(new bignumber_js_1.default(humanAmount).shiftedBy(decimals));
}
exports.humanTokenAmountToUint256 = humanTokenAmountToUint256;
function humanCollateralAmountToUint256(humanAmount) {
    return humanTokenAmountToUint256(humanAmount, types_1.BASE_DECIMALS);
}
exports.humanCollateralAmountToUint256 = humanCollateralAmountToUint256;
function humanEthAmountToUint256(humanAmount) {
    return humanTokenAmountToUint256(humanAmount, types_1.ETH_DECIMALS);
}
exports.humanEthAmountToUint256 = humanEthAmountToUint256;
// From-Contract Helpers
function uint256ToHumanEthAmount(output) {
    return uint256ToHumanTokenAmount(output, types_1.ETH_DECIMALS);
}
exports.uint256ToHumanEthAmount = uint256ToHumanEthAmount;
function uint256ToHumanCollateralTokenAmount(output) {
    return uint256ToHumanTokenAmount(output, types_1.BASE_DECIMALS);
}
exports.uint256ToHumanCollateralTokenAmount = uint256ToHumanCollateralTokenAmount;
function uint256ToHumanTokenAmount(output, decimals) {
    return new bignumber_js_1.default(output).shiftedBy(-decimals).toFixed();
}
exports.uint256ToHumanTokenAmount = uint256ToHumanTokenAmount;
