// @dydxprotocol/v3-client/build/src/modules/eth-private.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const starkex_lib_1 = require("@dydxprotocol/starkex-lib");
const lodash_1 = __importDefault(require("lodash"));
const eth_signing_1 = require("../eth-signing");
const request_helpers_1 = require("../helpers/request-helpers");
const axios_1 = require("../lib/axios");
const types_1 = require("../types");
class EthPrivate {
    constructor({ host, web3, networkId, clock, }) {
        this.host = host;
        this.signer = new eth_signing_1.SignEthPrivateAction(web3, networkId);
        this.clock = clock;
    }
    // ============ Request Helpers ============
    async request(method, endpoint, ethereumAddress, signingMethod, data = {}) {
        const requestPath = `/v3/${endpoint}`;
        const timestamp = this.clock.getAdjustedIsoString();
        const body = JSON.stringify(data);
        const signature = await this.signer.sign(ethereumAddress, signingMethod, {
            method,
            requestPath,
            body,
            timestamp,
        });
        return axios_1.axiosRequest({
            url: `${this.host}${requestPath}`,
            method,
            data: !lodash_1.default.isEmpty(data) ? body : undefined,
            headers: {
                'DYDX-SIGNATURE': signature,
                'DYDX-TIMESTAMP': timestamp,
                'DYDX-ETHEREUM-ADDRESS': ethereumAddress,
            },
        });
    }
    async post(endpoint, ethereumAddress, signingMethod = types_1.SigningMethod.Hash) {
        return this.request(starkex_lib_1.ApiMethod.POST, endpoint, ethereumAddress, signingMethod);
    }
    async delete(endpoint, ethereumAddress, signingMethod = types_1.SigningMethod.Hash, params) {
        const requestPath = request_helpers_1.generateQueryPath(endpoint, params);
        return this.request(starkex_lib_1.ApiMethod.DELETE, requestPath, ethereumAddress, signingMethod);
    }
    async get(endpoint, ethereumAddress, signingMethod = types_1.SigningMethod.Hash) {
        return this.request(starkex_lib_1.ApiMethod.GET, endpoint, ethereumAddress, signingMethod);
    }
    // ============ Requests ============
    /**
     * @description have an auto-generated apikey, secret and passphrase generated
     * for an ethereumAddress.
     * @param ethereumAddress the apiKey will be for
     * @param signingMethod used to validate the request
     */
    async createApiKey(ethereumAddress, signingMethod = types_1.SigningMethod.Hash) {
        return this.post('api-keys', ethereumAddress, signingMethod);
    }
    /**
     *
     * @param apiKey to be deleted
     * @param ethereumAddress the apiKey is for
     * @param signingMethod used to validate the request
     */
    async deleteApiKey(apiKey, ethereumAddress, signingMethod = types_1.SigningMethod.Hash) {
        return this.delete('api-keys', ethereumAddress, signingMethod, { apiKey });
    }
    /**
     * @description This is for if you can't recover your starkKey or apiKey and need an
     * additional way to get your starkKey, positionid and balance on our exchange,
     *  all of which are needed to call the L1 solidity function needed to recover your funds.
     *
     * @param ethereumAddress the recovery is for
     * @param signingMethod used to validate the request
     */
    async recovery(ethereumAddress, signingMethod = types_1.SigningMethod.Hash) {
        return this.get('recovery', ethereumAddress, signingMethod);
    }
}
exports.default = EthPrivate;
