// @dydxprotocol/v3-client/build/src/modules/public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_helpers_1 = require("../helpers/request-helpers");
const axios_1 = require("../lib/axios");
class Public {
    constructor(host) {
        this.host = host;
    }
    // ============ Request Helpers ============
    get(requestPath, params) {
        return axios_1.axiosRequest({
            method: 'GET',
            url: `${this.host}/v3/${request_helpers_1.generateQueryPath(requestPath, params)}`,
        });
    }
    async put(requestPath, data) {
        return axios_1.axiosRequest({
            url: `${this.host}/v3/${requestPath}`,
            method: 'PUT',
            data,
        });
    }
    // ============ Requests ============
    /**
     * @description check if a user exists for an ethereum address
     *
     * @param ethereumAddress of the user
     */
    doesUserExistWithAddress(ethereumAddress) {
        const uri = 'users/exists';
        return this.get(uri, { ethereumAddress });
    }
    /**
     * @description check if a username already exists
     *
     * @param username being queried
     */
    doesUserExistWithUsername(username) {
        const uri = 'usernames';
        return this.get(uri, { username });
    }
    /**
     * @description get market information for either all markets or a specific market
     *
     * @param market if only one market should be returned
     */
    getMarkets(market) {
        const uri = 'markets';
        return this.get(uri, { market });
    }
    /**
     * @description get orderbook for a specific market
     *
     * @param market being queried
     */
    getOrderBook(market) {
        return this.get(`orderbook/${market}`, {});
    }
    /**
     * @description get one or more market specific statistics for a time period
     *
     * @param {
     * @market being queried
     * @days if a specific time period statistic should be returned
     * }
     */
    getStats({ market, days, }) {
        const uri = market !== undefined
            ? `stats/${market}`
            : 'stats';
        return this.get(uri, { days });
    }
    /**
     * @description get trades for a market up to a certain time
     *
     * @param market being checked
     * @param startingBeforeOrAt latest trade being returned
     */
    getTrades({ market, startingBeforeOrAt, }) {
        const uri = `trades/${market}`;
        return this.get(uri, { startingBeforeOrAt });
    }
    /**
     * @description get historical funding rates for a market up to a certain time
     *
     * @param market being checked
     * @param effectiveBeforeOrAt latest historical funding rate being returned
     */
    getHistoricalFunding({ market, effectiveBeforeOrAt, }) {
        const uri = `historical-funding/${market}`;
        return this.get(uri, { effectiveBeforeOrAt });
    }
    /**
     * @description Get the amount of funds available for fast withdrawals, denominated in USDC.
     * To request a quote for a fast withdrawal, provide either a creditAmount or debitAmount (but
     * not both), and a creditAsset.
     *
     * @param creditAsset The asset to receive
     * @param creditAmount The amount to receive
     * @param debitAmount The amount of the collateral asset to transfer to the LP on layer-2
     */
    getFastWithdrawals({ creditAsset, creditAmount, debitAmount, }) {
        return this.get('fast-withdrawals', { creditAsset, creditAmount, debitAmount });
    }
    /**
     * @description get candles for a specific market
     *
     * @param market being checked
     * @param resolution Specific candle resolution being returned
     * @param fromISO is starting time candles are from
     * @param toISO is ending time candles go up to
     * @param limit to number of candles returned
     */
    getCandles({ market, resolution, fromISO, toISO, limit, }) {
        const uri = `candles/${market}`;
        return this.get(uri, {
            resolution,
            fromISO,
            toISO,
            limit,
        });
    }
    /**
     * @description get leaderboard pnls
     *
     * @param period Time period being checked
     * @param sortBy Pnl to sort by
     * @param limit Number of leaderboard pnls returned
     */
    getLeaderboardPnls({ period, startingBeforeOrAt, sortBy, limit, }) {
        const uri = 'leaderboard-pnl';
        return this.get(uri, {
            period,
            startingBeforeOrAt,
            sortBy,
            limit,
        });
    }
    /**
     * @description get retroactive mining rewards for an ethereum address
     *
     * @param ethereumAddress An Ethereum address of a user
     */
    getPublicRetroactiveMiningRewards(ethereumAddress) {
        const uri = 'rewards/public-retroactive-mining';
        return this.get(uri, {
            ethereumAddress,
        });
    }
    /**
     * @description verify email for user with token
     *
     * @token that verifies user received a verification email to
     * the email they specified
     */
    async verifyEmail(token) {
        return this.put('emails/verify-email', {
            token,
        });
    }
    /**
     * @description get currently revealed Hedgies
     */
    getCurrentlyRevealedHedgies() {
        return this.get('hedgies/current', {});
    }
    /**
     * @description get historically revealed Hedgies
     */
    getHistoricallyRevealedHedgies({ nftRevealType, start, end, }) {
        return this.get('hedgies/history', {
            nftRevealType,
            start,
            end,
        });
    }
    /**
     * @description get api server time as iso and as epoch in seconds with MS
     */
    getTime() {
        return this.get('time', {});
    }
    /**
     * @description get a rough estimate of the difference (in epoch seconds) between the server time
     * and the system time.
     */
    async getTimestampAdjustment() {
        const time1 = Date.now();
        const { epoch } = await this.getTime();
        const time2 = Date.now();
        const averageEpoch = (time1 + time2) / 2 / 1000;
        return epoch - averageEpoch;
    }
    /**
     * @description  get global config variables for the exchange as a whole.
     * This includes (but is not limited to) details on the exchange, including addresses,
     * fees, transfers, and rate limits.
     */
    async getConfig() {
        return this.get('config', {});
    }
}
exports.default = Public;
