// @dydxprotocol/starkex-lib/build/__tests__/helpers/assets.test.js
"use strict";
/**
 * Unit tests for helpers/assets.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const constants_1 = require("../../src/constants");
const types_1 = require("../../src/types");
// Module under test.
const assets_1 = require("../../src/helpers/assets");
describe('assets helpers', () => {
    describe('fromQuantums()', () => {
        it('converts a number of quantums to a human-readable amount', () => {
            (0, expect_1.default)((0, assets_1.fromQuantums)('1000', types_1.DydxAsset.ETH)).toBe('0.000001');
        });
        it('throws if the asset is unknown', () => {
            (0, expect_1.default)(() => {
                (0, assets_1.fromQuantums)('1000', 'UNKNOWN');
            }).toThrow('Unknown asset');
        });
    });
    describe('getStarkwareAmounts()', () => {
        it('converts order params to Starkware order params', () => {
            (0, expect_1.default)((0, assets_1.getStarkwareAmounts)({
                market: types_1.DydxMarket.BTC_USD,
                side: types_1.StarkwareOrderSide.SELL,
                humanSize: '250.0000000001',
                humanPrice: '1.23456789',
            }, types_1.NetworkId.ROPSTEN)).toStrictEqual({
                quantumsAmountSynthetic: '2500000000001',
                quantumsAmountCollateral: '308641972',
                assetIdSynthetic: constants_1.SYNTHETIC_ASSET_ID_MAP[types_1.DydxAsset.BTC],
                assetIdCollateral: constants_1.COLLATERAL_ASSET_ID_BY_NETWORK_ID[types_1.NetworkId.ROPSTEN],
                isBuyingSynthetic: false,
            });
        });
        it('converts order params with a quote amount instead of price', () => {
            (0, expect_1.default)((0, assets_1.getStarkwareAmounts)({
                market: types_1.DydxMarket.BTC_USD,
                side: types_1.StarkwareOrderSide.SELL,
                humanSize: '250.0000000001',
                humanQuoteAmount: '308.641972',
            }, types_1.NetworkId.ROPSTEN)).toStrictEqual({
                quantumsAmountSynthetic: '2500000000001',
                quantumsAmountCollateral: '308641972',
                assetIdSynthetic: constants_1.SYNTHETIC_ASSET_ID_MAP[types_1.DydxAsset.BTC],
                assetIdCollateral: constants_1.COLLATERAL_ASSET_ID_BY_NETWORK_ID[types_1.NetworkId.ROPSTEN],
                isBuyingSynthetic: false,
            });
        });
        it('throws if the order size is not a multiple of the Starkware quantum', () => {
            (0, expect_1.default)(() => {
                (0, assets_1.getStarkwareAmounts)({
                    market: types_1.DydxMarket.BTC_USD,
                    side: types_1.StarkwareOrderSide.SELL,
                    humanSize: '250.00000000001',
                    humanPrice: '1.23456789',
                }, types_1.NetworkId.ROPSTEN);
            }).toThrow('not a multiple of the quantum size');
        });
        it('throws if the quote amount is given and is not a multiple of the Starkware quantum', () => {
            (0, expect_1.default)(() => {
                (0, assets_1.getStarkwareAmounts)({
                    market: types_1.DydxMarket.BTC_USD,
                    side: types_1.StarkwareOrderSide.SELL,
                    humanSize: '250.0000000001',
                    humanQuoteAmount: '308.6419721',
                }, types_1.NetworkId.ROPSTEN);
            }).toThrow('not a multiple of the quantum size');
        });
    });
    describe('toQuantumsExact()', () => {
        it('converts a human readable amount to an integer number of quantums', () => {
            (0, expect_1.default)((0, assets_1.toQuantumsExact)('12.0000003', types_1.DydxAsset.LINK)).toBe('120000003');
        });
        it('throws if the amount does not divide evenly by the quantum size', () => {
            (0, expect_1.default)(() => {
                (0, assets_1.toQuantumsExact)('12.00000031', types_1.DydxAsset.LINK);
            }).toThrow('not a multiple of the quantum size');
        });
    });
    describe('toQuantumsRoundDown()', () => {
        it('converts a human readable amount to an integer number of quantums', () => {
            (0, expect_1.default)((0, assets_1.toQuantumsRoundDown)('12.0000003', types_1.DydxAsset.LINK)).toBe('120000003');
        });
        it('rounds down if the amount does not divide evenly by the quantum size', () => {
            (0, expect_1.default)((0, assets_1.toQuantumsRoundDown)('12.00000031', types_1.DydxAsset.LINK)).toBe('120000003');
        });
    });
    describe('toQuantumsRoundUp()', () => {
        it('converts a human readable amount to an integer number of quantums', () => {
            (0, expect_1.default)((0, assets_1.toQuantumsRoundUp)('12.0000003', types_1.DydxAsset.LINK)).toBe('120000003');
        });
        it('rounds up if the amount does not divide evenly by the quantum size', () => {
            (0, expect_1.default)((0, assets_1.toQuantumsRoundUp)('12.00000031', types_1.DydxAsset.LINK)).toBe('120000004');
        });
    });
    describe('getStarkwareLimitFeeAmount()', () => {
        it('converts the order limit fee as expected (edge case)', () => {
            (0, expect_1.default)((0, assets_1.getStarkwareLimitFeeAmount)('0.000001999999999999999999999999999999999999999999', '50750272151')).toBe('50751');
        });
    });
});
