// @dydxprotocol/v3-client/build/src/dydx-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DydxClient = void 0;
const starkex_eth_1 = require("@dydxprotocol/starkex-eth");
const web3_1 = __importDefault(require("web3"));
const clock_1 = __importDefault(require("./modules/clock"));
const eth_private_1 = __importDefault(require("./modules/eth-private"));
const onboarding_1 = __importDefault(require("./modules/onboarding"));
const private_1 = __importDefault(require("./modules/private"));
const public_1 = __importDefault(require("./modules/public"));
class DydxClient {
    constructor(host, options = {}) {
        this.host = host;
        this.apiTimeout = options.apiTimeout;
        this.ethSendOptions = options.ethSendOptions;
        this.networkId = typeof options.networkId === 'number' ? options.networkId : 1;
        this.starkPrivateKey = options.starkPrivateKey;
        this.apiKeyCredentials = options.apiKeyCredentials;
        if (options.web3 || options.web3Provider) {
            // Non-null assertion is safe due to if-condition.
            this.web3 = options.web3 || new web3_1.default(options.web3Provider);
        }
        // Modules.
        this._public = new public_1.default(host);
        this._clock = new clock_1.default(options.timestampAdjustment);
    }
    /**
     * Get the public module, used for interacting with public endpoints.
     */
    get public() {
        return this._public;
    }
    /**
     * Get the clock module, used for adjusting system time to server time.
     */
    get clock() {
        return this._clock;
    }
    /**
     * Get the private module, used for interacting with endpoints that require API-key auth.
     */
    get private() {
        if (!this._private) {
            if (this.apiKeyCredentials) {
                this._private = new private_1.default({
                    host: this.host,
                    apiKeyCredentials: this.apiKeyCredentials,
                    starkPrivateKey: this.starkPrivateKey,
                    networkId: this.networkId,
                    clock: this._clock,
                });
            }
            else {
                return notSupported('Private endpoints are not supported since apiKeyCredentials was not provided');
            }
        }
        return this._private;
    }
    /**
     * Get the keys module, used for managing API keys. Requires Ethereum key auth.
     */
    get ethPrivate() {
        if (!this._ethPrivate) {
            if (this.web3) {
                this._ethPrivate = new eth_private_1.default({
                    host: this.host,
                    web3: this.web3,
                    networkId: this.networkId,
                    clock: this._clock,
                });
            }
            else {
                return notSupported('Eth private endpoints are not supported since neither web3 nor web3Provider was provided');
            }
        }
        return this._ethPrivate;
    }
    /**
     * Get the onboarding module, used to create a new user. Requires Ethereum key auth.
     */
    get onboarding() {
        if (!this._onboarding) {
            if (this.web3) {
                this._onboarding = new onboarding_1.default(this.host, this.web3, this.networkId);
            }
            else {
                return notSupported('Onboarding endpoints are not supported since neither web3 nor web3Provider was provided');
            }
        }
        return this._onboarding;
    }
    /**
     * Get the eth module, used for interacting with Ethereum smart contracts.
     */
    get eth() {
        if (!this._eth) {
            if (this.web3) {
                this._eth = new starkex_eth_1.StarkwareLib(this.web3.currentProvider, this.networkId, this.ethSendOptions);
            }
            else {
                return notSupported('Eth endpoints are not supported since neither web3 nor web3Provider was provided');
            }
        }
        return this._eth;
    }
}
exports.DydxClient = DydxClient;
/**
 * Returns a proxy object that throws with the given message when trying to call a function on it.
 */
function notSupported(errorMessage) {
    const handler = {
        get() {
            throw new Error(errorMessage);
        },
    };
    return new Proxy({}, handler);
}
