// @dydxprotocol/starkex-lib/build/__tests__/signable/hashes.test.js
"use strict";
/* eslint-disable @typescript-eslint/no-unused-expressions */
/**
 * Test caching of pedersen hashes.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const proxyquire_1 = __importDefault(require("proxyquire"));
const sinon_1 = __importDefault(require("sinon"));
const starkware_1 = require("../../src/lib/starkware");
const types_1 = require("../../src/types");
proxyquire_1.default.noPreserveCache();
// Mocks.
let mockPedersen;
let proxyquiredCrypto;
let proxyquiredHashes;
let mocks;
// Mock data.
const mockTransfer = {
    senderPositionId: '12345',
    receiverPositionId: '67890',
    receiverPublicKey: '05135ef87716b0faecec3ba672d145a6daad0aa46437c365d490022115aba674',
    humanAmount: '49.478023',
    expirationIsoTimestamp: '2020-09-17T04:15:55.028Z',
    clientId: 'This is an ID that the client came up with to describe this transfer',
};
const mockConditionalTransfer = {
    ...mockTransfer,
    factRegistryAddress: '0x12aa12aa12aa12aa12aa12aa12aa12aa12aa12aa',
    fact: '0x12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff12ff',
};
const mockOrder = {
    positionId: '12345',
    humanSize: '145.0005',
    limitFee: '0.125',
    market: types_1.DydxMarket.ETH_USD,
    side: types_1.StarkwareOrderSide.BUY,
    expirationIsoTimestamp: '2020-09-17T04:15:55.028Z',
    humanPrice: '350.00067',
    clientId: 'This is an ID that the client came up with to describe this order',
};
const mockWithdrawal = {
    positionId: '12345',
    humanAmount: '49.478023',
    expirationIsoTimestamp: '2020-09-17T04:15:55.028Z',
    clientId: 'This is an ID that the client came up with to describe this withdrawal',
};
describe('Pedersen hashes', () => {
    beforeEach(() => {
        // Reload the hashes module fresh each time, resetting the cache.
        mockPedersen = sinon_1.default.spy(starkware_1.pedersen);
        proxyquiredCrypto = (0, proxyquire_1.default)('../../src/lib/crypto/proxies', {
            '../starkware': {
                pedersen: mockPedersen,
            },
        });
        proxyquiredHashes = (0, proxyquire_1.default)('../../src/signable/hashes', {
            '../lib/crypto': proxyquiredCrypto,
        });
        mocks = {
            '../lib/crypto': proxyquiredCrypto,
            './hashes': proxyquiredHashes,
        };
    });
    it('conditional transfer: 5 hashes the first time, and 4 thereafter', async () => {
        const { SignableConditionalTransfer } = ((0, proxyquire_1.default)('../../src/signable/conditional-transfer', mocks));
        await SignableConditionalTransfer.fromTransfer(mockConditionalTransfer, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(5);
        // Expect fewer hashes the second time.
        mockPedersen.resetHistory();
        await SignableConditionalTransfer.fromTransfer(mockConditionalTransfer, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(4);
    });
    it('order: 4 hashes the first time, and 2 thereafter', async () => {
        const { SignableOrder } = (0, proxyquire_1.default)('../../src/signable/order', mocks);
        await SignableOrder.fromOrder(mockOrder, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(4);
        // Expect fewer hashes the second time.
        mockPedersen.resetHistory();
        await SignableOrder.fromOrder(mockOrder, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(2);
    });
    it('transfer: 4 hashes the first time, and 3 thereafter', async () => {
        const { SignableTransfer } = ((0, proxyquire_1.default)('../../src/signable/transfer', mocks));
        await SignableTransfer.fromTransfer(mockTransfer, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(4);
        // Expect fewer hashes the second time.
        mockPedersen.resetHistory();
        await SignableTransfer.fromTransfer(mockTransfer, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(3);
    });
    it('withdrawal: 1 hash the first time, and 1 thereafter', async () => {
        const { SignableWithdrawal } = (0, proxyquire_1.default)('../../src/signable/withdrawal', mocks);
        await SignableWithdrawal.fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN).getHash();
        (0, expect_1.default)(mockPedersen.callCount).toBe(1);
    });
    describe('after pre-computing hashes', () => {
        beforeEach(async () => {
            await proxyquiredHashes.preComputeHashes(types_1.NetworkId.ROPSTEN);
            mockPedersen.resetHistory();
        });
        it('conditional transfer: 4 hashes', async () => {
            const { SignableConditionalTransfer } = ((0, proxyquire_1.default)('../../src/signable/conditional-transfer', mocks));
            await SignableConditionalTransfer.fromTransfer(mockConditionalTransfer, types_1.NetworkId.ROPSTEN).getHash();
            (0, expect_1.default)(mockPedersen.callCount).toBe(4);
        });
        it('order: 2 hashes', async () => {
            const { SignableOrder } = (0, proxyquire_1.default)('../../src/signable/order', mocks);
            await SignableOrder.fromOrder(mockOrder, types_1.NetworkId.ROPSTEN).getHash();
            (0, expect_1.default)(mockPedersen.callCount).toBe(2);
        });
        it('conditional transfer: 3 hashes', async () => {
            const { SignableTransfer } = ((0, proxyquire_1.default)('../../src/signable/transfer', mocks));
            await SignableTransfer.fromTransfer(mockTransfer, types_1.NetworkId.ROPSTEN).getHash();
            (0, expect_1.default)(mockPedersen.callCount).toBe(3);
        });
        it('withdrawal: 1 hash', async () => {
            const { SignableWithdrawal } = (0, proxyquire_1.default)('../../src/signable/withdrawal', mocks);
            await SignableWithdrawal.fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN).getHash();
            (0, expect_1.default)(mockPedersen.callCount).toBe(1);
        });
    });
});
