// @dydxprotocol/starkex-lib/build/__tests__/signable/oracle-price.test.js
"use strict";
/**
 * Unit tests for signable/oracle-price.ts.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const lodash_1 = __importDefault(require("lodash"));
const types_1 = require("../../src/types");
const keys_1 = require("../../src/keys");
const util_1 = require("../util");
// Module under test.
const oracle_price_1 = require("../../src/signable/oracle-price");
const src_1 = require("../../src");
// Mock params.
const mockKeyPair = {
    publicKey: '1895a6a77ae14e7987b9cb51329a5adfb17bd8e7c638f92d6892d76e51cebcf',
    privateKey: '178047D3869489C055D7EA54C014FFB834A069C9595186ABE04EA4D1223A03F',
};
const mockOraclePrice = {
    market: types_1.DydxMarket.BTC_USD,
    oracleName: 'dYdX',
    humanPrice: '11512.34',
    isoTimestamp: '2020-01-01T00:00:00.000Z',
};
const mockSignature = ('020b64c5ead744a9a39bb20cee8193e15958d2f5bc065a3a31a8245d800907ae' +
    '0043e5681d7fd1e0720cc578e3d076ea29dbfe902f30445da8aa74bd112aa710');
describe('SignableOraclePrice', () => {
    describe('verifySignature()', () => {
        it('returns true for a valid signature', async () => {
            const result = await oracle_price_1.SignableOraclePrice
                .fromPriceWithMarket(mockOraclePrice, types_1.NetworkId.ROPSTEN)
                .verifySignature(mockSignature, mockKeyPair.publicKey);
            (0, expect_1.default)(result).toBe(true);
        });
        it('returns false for an invalid signature', async () => {
            // Mutate a single character in r.
            await Promise.all(lodash_1.default.range(1, 4).map(async (i) => {
                const badSignature = (0, util_1.mutateHexStringAt)(mockSignature, i);
                const result = await oracle_price_1.SignableOraclePrice
                    .fromPriceWithMarket(mockOraclePrice, types_1.NetworkId.ROPSTEN)
                    .verifySignature(badSignature, mockKeyPair.publicKey);
                (0, expect_1.default)(result).toBe(false);
            }));
            // Mutate a single character in s.
            await Promise.all(lodash_1.default.range(1, 4).map(async (i) => {
                const badSignature = (0, util_1.mutateHexStringAt)(mockSignature, i + 64);
                const result = await oracle_price_1.SignableOraclePrice
                    .fromPriceWithMarket(mockOraclePrice, types_1.NetworkId.ROPSTEN)
                    .verifySignature(badSignature, mockKeyPair.publicKey);
                (0, expect_1.default)(result).toBe(false);
            }));
        });
    });
    describe('sign()', () => {
        it('signs an oracle price, with a market', async () => {
            const signature = await oracle_price_1.SignableOraclePrice
                .fromPriceWithMarket(mockOraclePrice, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).toEqual(mockSignature);
        });
        it('signs an oracle price, with an asset name', async () => {
            const signature = await oracle_price_1.SignableOraclePrice
                .fromPriceWithAssetName({
                ...mockOraclePrice,
                assetName: (0, src_1.getSignedAssetName)(mockOraclePrice.market),
            }, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).toEqual(mockSignature);
        });
        it('generates a different signature when the market is different', async () => {
            const oraclePrice = {
                ...mockOraclePrice,
                market: types_1.DydxMarket.ETH_USD,
            };
            const signature = await oracle_price_1.SignableOraclePrice
                .fromPriceWithMarket(oraclePrice, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
        it('generates a different signature when the oracle name is different', async () => {
            const oraclePrice = {
                ...mockOraclePrice,
                oracleName: 'Other',
            };
            const signature = await oracle_price_1.SignableOraclePrice
                .fromPriceWithMarket(oraclePrice, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
        it('generates a different signature when the timestamp is different', async () => {
            const oraclePrice = {
                ...mockOraclePrice,
                isoTimestamp: new Date().toISOString(),
            };
            const signature = await oracle_price_1.SignableOraclePrice
                .fromPriceWithMarket(oraclePrice, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
        it('throws an error if the oracle name is too long', () => {
            const oraclePrice = {
                ...mockOraclePrice,
                oracleName: 'Other2',
            };
            (0, expect_1.default)(() => oracle_price_1.SignableOraclePrice.fromPriceWithMarket(oraclePrice, types_1.NetworkId.ROPSTEN)).toThrow('Input does not fit in numBits=40 bits');
        });
    });
    it('end-to-end', async () => {
        // Repeat some number of times.
        await Promise.all(lodash_1.default.range(3).map(async () => {
            const keyPair = (0, keys_1.generateKeyPairUnsafe)();
            const signableOraclePrice = oracle_price_1.SignableOraclePrice.fromPriceWithMarket(mockOraclePrice, types_1.NetworkId.ROPSTEN);
            const signature = await signableOraclePrice.sign(keyPair.privateKey);
            // Expect to be valid when verifying with the right public key.
            (0, expect_1.default)(await signableOraclePrice.verifySignature(signature, keyPair.publicKey)).toBe(true);
            // Expect to be invalid when verifying with a different public key.
            (0, expect_1.default)(await signableOraclePrice.verifySignature(signature, mockKeyPair.publicKey)).toBe(false);
        }));
    });
});
