// @dydxprotocol/starkex-lib/build/__tests__/signable/transfer.test.js
"use strict";
/**
 * Unit tests for signable/transfer.ts.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const lodash_1 = __importDefault(require("lodash"));
const types_1 = require("../../src/types");
const keys_1 = require("../../src/keys");
const util_1 = require("../util");
// Module under test.
const transfer_1 = require("../../src/signable/transfer");
// Mock params.
const mockKeyPair = {
    publicKey: '3b865a18323b8d147a12c556bfb1d502516c325b1477a23ba6c77af31f020fd',
    privateKey: '58c7d5a90b1776bde86ebac077e053ed85b0f7164f53b080304a531947f46e3',
};
const mockParams = {
    senderPositionId: '12345',
    receiverPositionId: '67890',
    receiverPublicKey: '05135ef87716b0faecec3ba672d145a6daad0aa46437c365d490022115aba674',
    humanAmount: '49.478023',
    expirationIsoTimestamp: '2020-09-17T04:15:55.028Z',
    clientId: 'This is an ID that the client came up with to describe this transfer',
};
const mockSignature = ('06b72146028a7f0092557a3a04e9916bd4ae1fba0a4bd92670ef80e2293f7386' +
    '04c0918a7a8e622e463d40f24984c23fd8bab2cd32980676ba666f55c6efeaf3');
describe('SignableTransfer', () => {
    describe('verifySignature()', () => {
        it('returns true for a valid signature', async () => {
            const result = await transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN)
                .verifySignature(mockSignature, mockKeyPair.publicKey);
            (0, expect_1.default)(result).toBe(true);
        });
        it('returns false for an invalid signature', async () => {
            // Mutate a single character in r.
            const badSignatureR = (0, util_1.mutateHexStringAt)(mockSignature, 1);
            const result1 = await transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN)
                .verifySignature(badSignatureR, mockKeyPair.publicKey);
            (0, expect_1.default)(result1).toBe(false);
            // Mutate a single character in s.
            const badSignatureS = (0, util_1.mutateHexStringAt)(mockSignature, 65);
            const result2 = await transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN)
                .verifySignature(badSignatureS, mockKeyPair.publicKey);
            (0, expect_1.default)(result2).toBe(false);
        });
    });
    describe('sign()', () => {
        it('signs a transfer', async () => {
            const signature = await transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN).sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).toEqual(mockSignature);
        });
        it('generates a different signature when the client ID is different', async () => {
            const transfer = {
                ...mockParams,
                clientId: `${mockParams.clientId}!`,
            };
            const signature = await transfer_1.SignableTransfer.fromTransfer(transfer, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
        it('generates a different signature when the receiver position ID is different', async () => {
            const transfer = {
                ...mockParams,
                receiverPositionId: (Number.parseInt(mockParams.receiverPositionId, 10) + 1).toString(),
            };
            const signature = await transfer_1.SignableTransfer.fromTransfer(transfer, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
    });
    describe('toStarkware()', () => {
        it('converts human amounts to quantum amounts and converts expiration to hours', () => {
            const starkwareTransfer = (transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN).toStarkware());
            (0, expect_1.default)(starkwareTransfer.quantumsAmount).toEqual('49478023');
            (0, expect_1.default)(starkwareTransfer.expirationEpochHours).toBe(444533);
        });
    });
    it('end-to-end', async () => {
        // Repeat some number of times.
        await Promise.all(lodash_1.default.range(3).map(async () => {
            const keyPair = (0, keys_1.generateKeyPairUnsafe)();
            const signable = transfer_1.SignableTransfer.fromTransfer(mockParams, types_1.NetworkId.ROPSTEN);
            const signature = await signable.sign(keyPair.privateKey);
            // Expect to be valid when verifying with the right public key.
            (0, expect_1.default)(await signable.verifySignature(signature, keyPair.publicKey)).toBe(true);
            // Expect to be invalid when verifying with a different public key.
            (0, expect_1.default)(await signable.verifySignature(signature, mockKeyPair.publicKey)).toBe(false);
        }));
    });
});
