// @dydxprotocol/starkex-lib/build/__tests__/signable/withdrawal.test.js
"use strict";
/**
 * Unit tests for signable/withdrawal.ts.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const lodash_1 = __importDefault(require("lodash"));
const types_1 = require("../../src/types");
const keys_1 = require("../../src/keys");
const helpers_1 = require("../../src/helpers");
const util_1 = require("../util");
// Module under test.
const withdrawal_1 = require("../../src/signable/withdrawal");
// Mock params.
const mockKeyPair = {
    publicKey: '3b865a18323b8d147a12c556bfb1d502516c325b1477a23ba6c77af31f020fd',
    privateKey: '58c7d5a90b1776bde86ebac077e053ed85b0f7164f53b080304a531947f46e3',
};
const mockWithdrawal = {
    positionId: '12345',
    humanAmount: '49.478023',
    expirationIsoTimestamp: '2020-09-17T04:15:55.028Z',
    clientId: 'This is an ID that the client came up with to describe this withdrawal',
};
const mockSignature = ('05e48c33f8205a5359c95f1bd7385c1c1f587e338a514298c07634c0b6c952ba' +
    '0687d6980502a5d7fa84ef6fdc00104db22c43c7fb83e88ca84f19faa9ee3de1');
describe('SignableWithdrawal', () => {
    describe('verifySignature()', () => {
        it('returns true for a valid signature', async () => {
            const result = await withdrawal_1.SignableWithdrawal
                .fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN)
                .verifySignature(mockSignature, mockKeyPair.publicKey);
            (0, expect_1.default)(result).toBe(true);
        });
        it('returns false for an invalid signature', async () => {
            // Mutate a single character in r.
            await Promise.all(lodash_1.default.range(1, 4).map(async (i) => {
                const badSignature = (0, util_1.mutateHexStringAt)(mockSignature, i);
                const result = await withdrawal_1.SignableWithdrawal
                    .fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN)
                    .verifySignature(badSignature, mockKeyPair.publicKey);
                (0, expect_1.default)(result).toBe(false);
            }));
            // Mutate a single character in s.
            await Promise.all(lodash_1.default.range(1, 4).map(async (i) => {
                const badSignature = (0, util_1.mutateHexStringAt)(mockSignature, i + 64);
                const result = await withdrawal_1.SignableWithdrawal
                    .fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN)
                    .verifySignature(badSignature, mockKeyPair.publicKey);
                (0, expect_1.default)(result).toBe(false);
            }));
        });
    });
    describe('sign()', () => {
        it('signs a withdrawal', async () => {
            const signature = await withdrawal_1.SignableWithdrawal
                .fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).toEqual(mockSignature);
        });
        it('signs a withdrawal with nonce instead of clientId', async () => {
            const withdrawalWithNonce = {
                ...mockWithdrawal,
                clientId: undefined,
                nonce: (0, helpers_1.nonceFromClientId)(mockWithdrawal.clientId),
            };
            const signature = await withdrawal_1.SignableWithdrawal
                .fromWithdrawalWithNonce(withdrawalWithNonce, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).toEqual(mockSignature);
        });
        it('generates a different signature when the client ID is different', async () => {
            const withdrawal = {
                ...mockWithdrawal,
                clientId: `${mockWithdrawal.clientId}!`,
            };
            const signature = await withdrawal_1.SignableWithdrawal
                .fromWithdrawal(withdrawal, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
        it('generates a different signature when the position ID is different', async () => {
            const withdrawal = {
                ...mockWithdrawal,
                positionId: (Number.parseInt(mockWithdrawal.positionId, 10) + 1).toString(),
            };
            const signature = await withdrawal_1.SignableWithdrawal
                .fromWithdrawal(withdrawal, types_1.NetworkId.ROPSTEN)
                .sign(mockKeyPair.privateKey);
            (0, expect_1.default)(signature).not.toEqual(mockSignature);
        });
    });
    describe('toStarkware()', () => {
        it('converts human amounts to quantum amounts and converts expiration to hours', () => {
            const starkwareWithdrawal = withdrawal_1.SignableWithdrawal
                .fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN)
                .toStarkware();
            (0, expect_1.default)(starkwareWithdrawal.quantumsAmount).toBe('49478023');
            (0, expect_1.default)(starkwareWithdrawal.expirationEpochHours).toBe(444533);
        });
    });
    it('end-to-end', async () => {
        // Repeat some number of times.
        await Promise.all(lodash_1.default.range(3).map(async () => {
            const keyPair = (0, keys_1.generateKeyPairUnsafe)();
            const signableWithdrawal = withdrawal_1.SignableWithdrawal.fromWithdrawal(mockWithdrawal, types_1.NetworkId.ROPSTEN);
            const signature = await signableWithdrawal.sign(keyPair.privateKey);
            // Expect to be valid when verifying with the right public key.
            (0, expect_1.default)(await signableWithdrawal.verifySignature(signature, keyPair.publicKey)).toBe(true);
            // Expect to be invalid when verifying with a different public key.
            (0, expect_1.default)(await signableWithdrawal.verifySignature(signature, mockKeyPair.publicKey)).toBe(false);
        }));
    });
});
