// @dydxprotocol/starkex-eth/build/src/StarkwareLib.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarkwareLib = void 0;
const web3_1 = __importDefault(require("web3"));
const Contracts_1 = require("./lib/Contracts");
const CollateralToken_1 = require("./modules/CollateralToken");
const Exchange_1 = require("./modules/Exchange");
const FactRegistry_1 = require("./modules/FactRegistry");
const Logs_1 = require("./modules/Logs");
const MintableToken_1 = require("./modules/MintableToken");
class StarkwareLib {
    constructor(provider, networkId, options) {
        this.web3 = new web3_1.default(provider);
        this.contracts = this.getContracts(provider, networkId, this.web3, options);
        this.collateralToken = new CollateralToken_1.CollateralToken(this.contracts);
        this.mintableToken = new MintableToken_1.MintableToken(this.contracts);
        this.exchange = new Exchange_1.Exchange(this.contracts);
        this.factRegistry = new FactRegistry_1.FactRegistry(this.contracts);
        this.logs = new Logs_1.Logs(this.web3, this.contracts);
    }
    setProvider(provider, networkId) {
        this.web3.setProvider(provider);
        this.contracts.setProvider(provider, networkId);
    }
    // ============ Helper Functions ============
    getContracts(provider, networkId, web3, options) {
        return new Contracts_1.Contracts(provider, networkId, web3, options);
    }
}
exports.StarkwareLib = StarkwareLib;
